File: selenium/src/java/org/openqa/selenium/internal/seleniumemulation/WaitForCondition.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium.internal.seleniumemulation;

import com.thoughtworks.selenium.Wait;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;

public class WaitForCondition extends SeleneseCommand<Void> {

  @Override
  protected Void handleSeleneseCommand(final WebDriver driver, final String script, final String timeout) {
    new Wait() {
      @Override
      public boolean until() {
        return (Boolean) ((JavascriptExecutor) driver).executeScript(script);
      }
    }.wait("Failed to resolve " + script, Long.valueOf(timeout));

    return null;
  }
}

File: selenium/src/csharp/WebdriverBackedSelenium/Internal/SeleniumEmulation/WaitForCondition.cs
using System;
using System.Collections.Generic;
using System.Text;
using OpenQA.Selenium;

namespace Selenium.Internal.SeleniumEmulation
{
    class WaitForCondition : SeleneseCommand
    {
        protected override object HandleSeleneseCommand(IWebDriver driver, string locator, string value)
        {
            ConditionWaiter waiter = new ConditionWaiter(driver, locator);
            waiter.Wait("Failed to resolve " + locator, long.Parse(value));
            return null;
        }

        private class ConditionWaiter : Waiter
        {
            private IWebDriver driver;
            private string script;

            public ConditionWaiter(IWebDriver driver, string script)
                : base()
            {
                this.driver = driver;
                this.script = script;
            }

            public override bool Until()
            {
                return (bool)((IJavaScriptExecutor)driver).ExecuteScript(script);
            }
        }
    }
}
